Removing dockershim from kubelet (#1985)
kubernetes/enhancements: keps/sig-node/1985-remove-dockershim/README.md

- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Terms

- **CRI:** Container Runtime Interface – a plugin interface which enables kubelet to use a wide variety of container 
runtimes, without the need to recompile.

## Summary

CRI for docker (i.e. dockershim) is currently a built-in container runtime in kubelet code base. This proposal aims 
at a deprecation and subsequent removal of dockershim from kubelet.

## Motivation

> In Kubernetes, CRI is the used as the "default" container runtime, while currently the CRI of docker (a.k.a. dockershim) 

Reviewer: typo? `CRI is the used as the "default" container runtime`

Author: Fixed

is part of kubelet code and coupled with kubelet's lifecycle. 

This is not ideal as kubelet then has dependency on specific container runtime which leads to maintenance burden for not 
only developers in sig-node, but also cluster administrators when critical issues (e.g. runc CVE) happen to container runtimes. The pros of removing dockershim is straightforward:

### Pros
- Docker is not special and should be just a CRI implementation just like every other CRI implementation in our ecosystem. 

> - Currently, dockershim "enjoys" some backdoors for various reasons (see [legacyLogProvider](https://cs.k8s.io/?q=legacyLogProvider&i=nope&files=&repos=kubernetes/kubernetes) for example) . Removing these "features" should eliminate maintenance burden of kubelet.

Reviewer: "backdoor" suggests security issues. possibly a term like "special privileges" or "inconsistent integrations"?

Author: Done. picked `inconsistent integrations`

- Docker is not special and should be just a CRI implementation just like every other CRI implementation in our ecosystem. 
- Currently, dockershim "enjoys" some inconsistent integrations for various reasons (see [legacyLogProvider](https://cs.k8s.io/?q=legacyLogProvider&i=nope&files=&repos=kubernetes/kubernetes) for example) . Removing these "features" should eliminate maintenance burden of kubelet.
- A cri-dockerd can be maintained independently by folks who are interested in keeping this functionality

> - Over time we can remove vendored docker dependencies in kubelet.

Reviewer 2: Another two pros in my mind:

1) Due to convenience inheriting from this builtin shim for the container runtime, there is less incentive to move to new container runtime. The production issues found and addressed in both Containerd and CRI-O might cause the production issues for some users, which causes the maintenance burden to Kubernetes community.
2) The community can focus on the new container runtimes enhancements.

Author: Included these 2 suggestions. thanks @dchen1107

### Cons
- Deployment pain with a new binary in addition to kubelet.
  - An additional component may aggravate the complexity currently. It may be relieved with docker version evolutions.

> - The number of affected users maybe large.

Reviewer: typo: may be

  - Users must change existing use experience when using Kubernetes and docker.
  - Users have to change their existing workflows to adapt to this new changes.
  - And other unrecorded stuff.

> - Updating all the eco-system tools to avoid dependencies on docker.

Author: Dropping this :)

- Many people use the built in dockershim for in-cluster image build. While that may not be something we recommend for 

> a variety of reasons, it will be a breaking change for these users.

Reviewer 3: should we recommend to those users to move to:
https://github.com/moby/buildkit
https://docs.docker.com/develop/develop-images/build_enhancements/
> Limitations
> Only supported for building Linux containers

Reviewer 4: They're not using **dockershim** for build. They're using dockerd directly. They could continue to run dockerd and use it outside of kubernetes. This usage isn't really relevant..?

Author: Dropping this too

> - CRI is still in alpha，should probably get a 1.0 out there splitting out dockershim completely from kubelet. There 

Reviewer: suggest strengthening this language:
```suggestion
- CRI is still technically in alpha, and should graduate to GA before removing dockershim from kubelet.
```

Author: Done

needs to be another KEP for graduating the CRI API version.

> - Existing CNI and CSI plugins may also be affected.

Reviewer: this surprised me... are there more concrete details you can point to for these?

Reviewer 5: The only CNI-related CRI-vs-dockershim difference that I know of is that CRI handles pod HostPorts itself, while dockershim expects someone else to do it. (Either the plugin does it itself or else the administrator configures a chain of plugins including `portmap`.) So a plugin that previously only supported dockershim might need to be updated to also support CRI correctly (eg, https://github.com/openshift/origin/pull/16463 though that PR was slightly wrong in that it assumed the difference was CRI-O-specific, not CRI in general). But all those issues probably got worked out years ago...

Author: Dropping this.

- kubelet has no dependency on dockershim/docker in its whole lifecycle. This is already done using the `dockerless` tag
- All node related features are CRI generic and have no "back door" dependency on dockershim/docker.
- Deprecate and remove, or replace all Docker-specific features.
- Reasonable benchmark result of performance degradation after moving dockershim to out-of-tree.
- E2E test framework has been updated with fully support of out-of-tree CRI container runtime.

### Dockershim removal plan

Step 1: Deprecate in-tree dockershim and decouple dockershim from kubelet.

Target releases: 1.20, 1.21

Actions:

- Mark in-tree dockershim as "maintenance mode":
  - CRI generic changes/features can continue on dockershim.
  - WIP efforts on dockershim can continue and go to complete.
  - dockershim/docker specific changes/features should be rejected.
- Deprecate the legacy features of dockershim in kubelet by providing a specific timeline. Currently, kubelet still has:
  - vendored dockershim 
  - flags that are used to configure dockershim.
  - support to get container logs when docker uses journald as the driver.
  - logic of moving docker processes to a given cgroup

>   - TBD anything else?

Reviewer: note: complete before merge

Author: Dropping this.

- Ensure e2e/Node e2e test framework is CRI generic and test cases are independent of container runtime.
- Refactoring e2e/Node e2e test framework to include CRI for docker installation (or use other CRI container runtime).
  - Ensure cluster/node e2e are 100% CRI focused.
  - Ensure test-infra install appropriate CRI implementations in e2e machines. 
- Ensure Windows scenarios that currently depend on docker are fully supported in alternative CRI implementations

Step 2: Release kubelet without dockershim

Target releases: 1.22

Actions:
- Document and announce migration guide.
- Release harness would build kubelet with `dockerless` tag on. So the default build will not support docker out of 
the box. 

> - If folks need this support, they would have to build kublet by themselves as the code is still present in the

Reviewer 6: ```suggestion
- If folks need this support, they would have to build kubelet by themselves as the code is still present in the
```

Author: Done

> source tree.

Reviewer: at first glance, this doesn't match the spirit of the deprecation policy, which is intended to avoid requiring consumers to change the way they consume/run kubernetes until the deprecation period ends

Author: Let's tweak this as we go forward as the KEP is a living document.

Step 3: Completely remove in-tree dockershim from kubelet.

Target releases: Deprecation should be for at least a year. So the earliest possible release after that time period.

Actions:

- Delete in-tree dockershim code from kubelet after certain "grace period".

### Risks and Mitigations

The easier we make it for folks to switch to CRI implementations the lesser the risk. Another option would be for
folks for a brand new CRI implementation that targets docker. Though even this option means that folks will have to
run an extra process outside of kubelet. The worst case scenario is for us to carry on the dockershim for a couple
of more releases. 

### Test Plan

Node e2e testing will be augmented to test kubelet built with `dockerless` tag 

### Graduation Criteria

- All feedback gathered from users
- Adequate test signal quality for node e2e
- Tests are in Testgrid and linked in KEP
- Allowing time for additional user feedback and bug reports

### Upgrade / Downgrade Strategy

> Not applicable.

Reviewer 3: maybe we should mention this here:

```suggestion
Upgrade: Users should follow the migration guide before upgrading to a version of the kubelet that no longer includes dockershim.

Downgrade: Not applicable.
```

Author: Updated

### Version Skew Strategy

Not applicable.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

_This section must be completed when targeting alpha to a release._

* **How can this feature be enabled / disabled in a live cluster?**
  - [ ] Feature gate (also fill in values in `kep.yaml`)

>     - Feature gate name: SizeMemoryBackedVolumes

Reviewer 7: This part is probably not relevant?

Author: cleaned up

Target releases: 1.20, 1.21

Actions:

- Mark in-tree dockershim as "maintenance mode":
  - CRI generic changes/features can continue on dockershim.
  - WIP efforts on dockershim can continue and go to complete.
  - dockershim/docker specific changes/features should be rejected.
- Deprecate the legacy features of dockershim in kubelet by providing a specific timeline. Currently, kubelet still has:
  - vendored dockershim 
  - flags that are used to configure dockershim.
  - support to get container logs when docker uses journald as the driver.
  - logic of moving docker processes to a given cgroup
- Ensure e2e/Node e2e test framework is CRI generic and test cases are independent of container runtime.
- Refactoring e2e/Node e2e test framework to include CRI for docker installation (or use other CRI container runtime).
  - Ensure cluster/node e2e are 100% CRI focused.
  - Ensure test-infra install appropriate CRI implementations in e2e machines. 
- Ensure Windows scenarios that currently depend on docker are fully supported in alternative CRI implementations

Step 2: Release kubelet without dockershim

Target releases: 1.22

Actions:

> - Document and announce migration guide.

Reviewer 8: is there a tracking issue for this migration guide, i think it would help to include in it what steps do existing users take for moving from docker to containerd running on there hosts ?

Reviewer 9: I started this: https://github.com/kubernetes/website/pull/25787
